### PR TITLE
Feature/#49 detail movie screen

### PR DESCRIPTION
**Pull request by IstrajI**

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -26,6 +26,10 @@ android {
         sourceCompatibility versions.javaVersion
         targetCompatibility versions.javaVersion
     }
+
+    dexOptions {
+        preDexLibraries = false
```
Comment (mKantsavoi) [inline app/build.gradle]: We still need it?

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -1,30 +1,37 @@
...
-                <action android:name="android.intent.action.MAIN"/>
-                <category android:name="android.intent.category.LAUNCHER"/>
+                <action android:name="android.intent.action.MAIN" />
+                <category android:name="android.intent.category.LAUNCHER" />
             </intent-filter>
         </activity>
 
         <activity
             android:name=".ui.activities.home.HomeActivity"
+            android:theme="@style/PopcornBaseTheme" />
```
Comment (mKantsavoi) [inline app/src/main/AndroidManifest.xml]: we have global app theme. Remove this pls

File: app/src/main/java/com/blaizmiko/popcornapp/presentation/presenters/movies/DetailMoviesPresenter.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import com.blaizmiko.popcornapp.presentation.views.movies.MovieDetailsView;
+
+import javax.inject.Inject;
+
+import rx.Subscription;
+import rx.android.schedulers.AndroidSchedulers;
+import rx.schedulers.Schedulers;
+
+@InjectViewState
+public class DetailMoviesPresenter extends BaseMvpPresenter<MovieDetailsView>{
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/presentation/presenters/movies/DetailMoviesPresenter.java]: if u rename activity and view, mb better will be rename presenter too

File: app/src/main/java/com/blaizmiko/popcornapp/presentation/presenters/movies/NowMoviesPresenter.java [hunk 4]
```diff
@@ -33,7 +33,7 @@ public void loadNowMoviesList() {
                 .getNowPlayingMovies(Constants.Api.ApiKey, Constants.Api.Language, mCurrentPage, Constants.Api.NowMovieDefaultRegion)
                 .flatMap(nowPlayingMovies -> Observable.from(nowPlayingMovies.getMovies()))
                 .filter(briefMovie -> briefMovie != null)
-                .map(briefMovie -> new TileAdapter.Item(briefMovie.getBackdropPath(), briefMovie.getTitle(), briefMovie.getVoteAverage()))
+                .map(briefMovie -> new TileAdapter.Item(briefMovie.getBackdropPath(), briefMovie.getTitle(), briefMovie.getVoteAverage(), briefMovie.getId()))
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/presentation/presenters/movies/NowMoviesPresenter.java]: not error, but better set id first in constructor

File: app/src/main/java/com/blaizmiko/popcornapp/ui/ActivityNavigator.java [hunk 5]
```diff
@@ -13,4 +14,10 @@ public static void startHomeActivity(@NonNull final Context context) {
         final Intent intent = new Intent(context, HomeActivity.class);
         context.startActivity(intent);
     }
+
+    public static void startMovieDetailsActivity(@NonNull final Context context, final int elementId) {
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/ui/ActivityNavigator.java]: elementId to id

File: app/src/main/java/com/blaizmiko/popcornapp/ui/ActivityNavigator.java [hunk 6]
```diff
@@ -13,4 +14,10 @@ public static void startHomeActivity(@NonNull final Context context) {
...
+        final Intent intent = new Intent(context, MovieDetailsActivity.class);
+        intent.putExtra("ID", elementId);
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/ui/ActivityNavigator.java]: send ID variable to constants with special name interface

File: app/src/main/java/com/blaizmiko/popcornapp/ui/activities/movie/MovieDetailsActivity.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+import butterknife.BindView;
+
+public class MovieDetailsActivity extends BaseActivity {
+
+    public static MovieDetailsFragment newInstance() {
+        return new MovieDetailsFragment();
+    }
+
+    private final int mDefaultMovieDetailsId = 0;
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/ui/activities/movie/MovieDetailsActivity.java]: why global variable?

File: app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java [hunk 8]
```diff
@@ -30,11 +32,13 @@
     private final Context mContext;
     private final List<Item> mItems;
     private final TileType mTileType;
+    private OnClickShowDetailsListener mItemClickListener;
 
-    public TileAdapter(final Context context, final TileType tileType) {
+    public TileAdapter(final Context context, final TileType tileType, OnClickShowDetailsListener listener) {
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java]: yeah, we need this, but set with constructor is wrong way

File: app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java [hunk 9]
```diff
@@ -61,6 +65,8 @@ public void onBindViewHolder(final TileAdapter.ViewHolder holder, final int posi
                 .load(mItems.get(position).getImageUrl())
                 .diskCacheStrategy(DiskCacheStrategy.SOURCE)
                 .into(holder.posterImageView);
+
+        holder.bind(mItems.get(position).getId());
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java]: bad solution

File: app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java [hunk 10]
```diff
@@ -86,6 +92,15 @@ public int getItemCount() {
             super(view);
             ButterKnife.bind(this, view);
         }
+
+        public void bind(int contentId) {
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java]: bad solution

File: app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java [hunk 11]
```diff
@@ -86,6 +92,15 @@ public int getItemCount() {
...
+            itemView.setOnClickListener(new View.OnClickListener() {
+                @Override
+                public void onClick(View v) {
+                    mItemClickListener.onClick(contentId);
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java]: bad solution

File: app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java [hunk 12]
```diff
@@ -132,4 +153,8 @@ String getRatingAsString() {
     public enum TileType {
         VERTICAL_TILE, HORIZONTAL_TILE
     }
+
+    public interface OnClickShowDetailsListener {
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java]: bad solution

File: app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/movies/GenresTagsAdapter.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+    @Override
+    public ViewHolder onCreateViewHolder(final ViewGroup parent, final int viewType) {
+        final View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.adapter_detail_movies_genre_tag_item, parent, false);
+        return new ViewHolder(view);
+    }
+
+    @Override
+    public void onBindViewHolder(final ViewHolder holder, int position) {
+        holder.setIsRecyclable(false);
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/movies/GenresTagsAdapter.java]: what it is?

File: app/src/main/java/com/blaizmiko/popcornapp/ui/fragments/movies/MovieDetailsFragment.java [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+
+    @InjectPresenter
+    LoadProgressPresenter mLoadProgressPresenter;
+
+    private GenresTagsAdapter mGenresTagsAdapter;
+    private int mMovieId;
+    private final int mStoryLineTextViewLinesMin = 3;
+    private final int mStoryLineTextViewLinesMax = 8;
+    private boolean mIsStoryLineTextViewOpen = false;
+    public static final String mBundleArgumentId = "ID";
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/ui/fragments/movies/MovieDetailsFragment.java]: to constants

File: app/src/main/java/com/blaizmiko/popcornapp/ui/fragments/movies/MovieDetailsFragment.java [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+    TextView mRatingTextView;
+    @BindView(R.id.fragment_detail_movie_title_text_view)
+    TextView mTitleTextView;
+    @BindView(R.id.fragment_detail_movie_story_line_text_view)
+    TextView mStoryLineTextView;
+    @BindView(R.id.fragment_detail_movie_rating_bar)
+    SimpleRatingBar mRatingBar;
+    @BindView(R.id.fragment_detail_movie_genre_tags_recycler_view)
+    RecyclerView mGenresRecyclerView;
+    @BindView(R.id.fragment_detail_movie_short_info_linear_layout)
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/ui/fragments/movies/MovieDetailsFragment.java]: info_layout will be better. W dont care about layout type

File: app/src/main/java/com/blaizmiko/popcornapp/ui/fragments/movies/MovieDetailsFragment.java [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+        mGenresTagsAdapter = new GenresTagsAdapter();
+        mGenresRecyclerView.setAdapter(mGenresTagsAdapter);
+    }
+
+    @Override
+    public void setMovie(Movie movie) {
+        mRatingTextView.setText(Double.toString(movie.getVoteAverage()));
+        mTitleTextView.setText(movie.getTitle());
+        mStoryLineTextView.setText(movie.getOverview());
+        mRatingBar.setRating(Float.parseFloat(new DecimalFormat(FormatUtils.ONE_DECIMAL, new DecimalFormatSymbols(Locale.US)).format(movie.getVoteAverage() / 2)));
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/ui/fragments/movies/MovieDetailsFragment.java]: so, mb we need separate method in special class for this, and after that reuse this method

File: app/src/main/res/layout/fragment_detail_movies.xml [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+            android:layout_width="match_parent"
+            android:layout_height="match_parent">
+
+    <FrameLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content">
+
+        <ProgressBar
+            android:id="@+id/fragment_detail_movie_progress_bar"
+            style="@style/Base.Widget.AppCompat.ProgressBar.Horizontal"
```
Comment (mKantsavoi) [inline app/src/main/res/layout/fragment_detail_movies.xml]: make special style for this

File: ui/src/main/res/layout/transparent_toolbar.xml [hunk 18]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<android.support.v7.widget.Toolbar
+    android:id="@+id/toolbar"
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="?attr/actionBarSize"
+    android:background="?attr/colorTransparent"
```
Comment (mKantsavoi) [inline ui/src/main/res/layout/transparent_toolbar.xml]: its funny

File: ui/src/main/res/values/dimen.xml [hunk 19]
```diff
@@ -27,12 +27,16 @@
     <dimen name="image_tile_vertical_width">98dp</dimen>
     <dimen name="image_tile_vertical_height">140dp</dimen>
 
+    <dimen name="fragment_detail_movies_top_spacing">100dp</dimen>
```
Comment (mKantsavoi) [inline ui/src/main/res/values/dimen.xml]: bad solution and name

File: ui/src/main/res/values/dimen.xml [hunk 20]
```diff
@@ -27,12 +27,16 @@
...
+    <dimen name="fragment_detail_movie_genre_tags_spacing">60dp</dimen>
```
Comment (mKantsavoi) [inline ui/src/main/res/values/dimen.xml]: 60dp - spacing?

File: app/src/main/res/layout/fragment_detail_movies.xml [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+                        android:id="@+id/fragment_detail_movie_poster_image_view"
+                        android:layout_width="@dimen/image_tile_vertical_width"
+                        android:layout_height="@dimen/image_tile_vertical_height"
+                        android:layout_margin="@dimen/spacing_1"
+                        android:adjustViewBounds="true"
+                        android:scaleType="centerCrop"/>
+                </FrameLayout>
+
+                <LinearLayout
+                    android:id="@+id/fragment_detail_movie_short_info_linear_layout"
```
Comment (mKantsavoi) [inline app/src/main/res/layout/fragment_detail_movies.xml]: without linear in name

File: ui/src/main/res/layout/transparent_toolbar.xml (hunk 18 above)
Comment (IstrajI) [reply ui/src/main/res/layout/transparent_toolbar.xml]: What? Dont understand.

Comment (mKantsavoi) [reply ui/src/main/res/layout/transparent_toolbar.xml]: Bad solution =)

File: ui/src/main/res/values/dimen.xml (hunk 19 above)
Comment (IstrajI) [reply ui/src/main/res/values/dimen.xml]: Should I use percents for Backdrop image? And percents for block with poster and info?

File: ui/src/main/res/values/dimen.xml (hunk 20 above)
Comment (IstrajI) [reply ui/src/main/res/values/dimen.xml]: Hole?

File: ui/src/main/res/values/dimen.xml (hunk 19 above)
Comment (mKantsavoi) [reply ui/src/main/res/values/dimen.xml]: We need to achieve a similar effect without the use of specific values

File: ui/src/main/res/values/dimen.xml (hunk 20 above)
Comment (mKantsavoi) [reply ui/src/main/res/values/dimen.xml]: between what?

File: app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java (hunk 11 above)
Comment (IstrajI) [reply app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java]: Should I create class in ui-listeners for this?

Comment (mKantsavoi) [reply app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java]: nope, inside adapter

File: app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java (hunk 12 above)
Comment (IstrajI) [reply app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java]: this

File: app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java (hunk 8 above)
Comment (IstrajI) [reply app/src/main/java/com/blaizmiko/popcornapp/ui/adapters/TileAdapter.java]: this

File: app/src/main/java/com/blaizmiko/popcornapp/ui/activities/base/BaseActivity.java [hunk 22]
```diff
@@ -54,7 +54,8 @@ protected void setToolbar(@NonNull final Toolbar toolbar) {
 
     protected void setToolbarTitle(final CharSequence title) {
         final ActionBar actionBar = getSupportActionBar();
-
+        System.out.println("setting");
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/ui/activities/base/BaseActivity.java]: !!!!!!

File: app/src/main/java/com/blaizmiko/popcornapp/presentation/presenters/movies/MovieDetailsPresenter.java [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+                    loadRating(movie.getImdbId());
+                }, error -> {
+                    getViewState().finishLoad();
+                    getViewState().showError();
+                }, () -> getViewState().finishLoad());
+
+        unSubscribeOnDestroy(detailMovieSubscription);
+    }
+
+    public void loadRating(String id) {
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/presentation/presenters/movies/MovieDetailsPresenter.java]: 1) Move to separate presenter
2) Or make parallel loading with the movie

File: app/src/main/java/com/blaizmiko/popcornapp/ui/activities/movie/MovieDetailsActivity.java [hunk 24]
```diff
@@ -0,0 +1,221 @@
...
+        Glide.with(getApplicationContext())
+                .load(Constants.TheMovieDbApi.BaseHighResImageUrl + movie.getBackdropPath())
+                .diskCacheStrategy(DiskCacheStrategy.SOURCE)
+                .into(mBackdropImageView);
+        Glide.with(getApplicationContext())
+                .load(Constants.TheMovieDbApi.BaseLowResImageUrl + movie.getPosterPath())
+                .diskCacheStrategy(DiskCacheStrategy.SOURCE)
+                .into(mPosterImageView);
+
+        mGenresTagsAdapter.update(movie.getGenres());
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/ui/activities/movie/MovieDetailsActivity.java]: this and below can be separate methods

File: app/src/main/java/com/blaizmiko/popcornapp/ui/activities/movie/MovieDetailsActivity.java [hunk 25]
```diff
@@ -0,0 +1,221 @@
...
+
+    @Override
+    public void hideProgress() {
+        if (mProgressBar.getVisibility() != View.GONE) {
+            mProgressBar.setVisibility(View.GONE);
+        }
+    }
+
+    @Override
+    public void onClick(View v) {
```
Comment (mKantsavoi) [inline app/src/main/java/com/blaizmiko/popcornapp/ui/activities/movie/MovieDetailsActivity.java]: Do not mix listeners and views callbacks

File: ui/src/main/res/layout/details_toolbar.xml [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+        app:expandedTitleTextAppearance="@style/Style.Toolbar.TransparentTitle"
+        app:layout_scrollFlags="scroll|exitUntilCollapsed">
+
+        <LinearLayout
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:orientation="vertical">
+
+            <ImageView
+                android:id="@+id/fragment_detail_movie_backdrop_image_view"
```
Comment (mKantsavoi) [inline ui/src/main/res/layout/details_toolbar.xml]: Wrong name for separate toolbar

same below

File: ui/src/main/res/values/styles.xml [hunk 27]
```diff
@@ -149,4 +149,11 @@
     <style name="Style.Button.Flat" parent="Widget.AppCompat.Button.Borderless.Colored">
         <item name="android:textAppearance">?attr/textAppearance.Button</item>
     </style>
+
+    <style name="Style.ProgressBar.Horizontal" parent="Base.Widget.AppCompat.ProgressBar.Horizontal"/>
+
+    <style name="Style.Toolbar.TransparentTitle" parent="ThemeOverlay.AppCompat.Dark.ActionBar">
```
Comment (mKantsavoi) [inline ui/src/main/res/values/styles.xml]: ?

File: ui/src/main/res/values/styles.xml [hunk 28]
```diff
@@ -149,4 +149,11 @@
     <style name="Style.Button.Flat" parent="Widget.AppCompat.Button.Borderless.Colored">
         <item name="android:textAppearance">?attr/textAppearance.Button</item>
     </style>
+
+    <style name="Style.ProgressBar.Horizontal" parent="Base.Widget.AppCompat.ProgressBar.Horizontal"/>
```
Comment (mKantsavoi) [inline ui/src/main/res/values/styles.xml]: So bring here the basic elements of style

File: app/src/main/res/layout/activity_detail_movies.xml [hunk 29]
```diff
@@ -0,0 +1,185 @@
...
+        android:layout_height="match_parent"
+        app:layout_behavior="@string/appbar_scrolling_view_behavior">
+
+        <LinearLayout
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:orientation="vertical">
+
+            <ProgressBar
+                android:id="@+id/fragment_detail_movie_progress_bar"
```
Comment (mKantsavoi) [inline app/src/main/res/layout/activity_detail_movies.xml]: its not fragment already, so wrong name

same below

**Comment by mKantsavoi** [comment]
And pls resolve conflicts

File: ui/src/main/res/values/styles.xml (hunk 27 above)
Comment (IstrajI) [reply ui/src/main/res/values/styles.xml]: I use it to disable title when toolbar is expanded. I could delete parent reference.

Comment (mKantsavoi) [reply ui/src/main/res/values/styles.xml]: hm, ok

**Pull request merged**